csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs

```diff
@@ -190,14 +190,470 @@ public IDisposableReadOnlyCollection<DisposableNamedOnnxValue> Run(IReadOnlyColl
+            IReadOnlyCollection<PinnedOnnxValue> inputValues,
+            IReadOnlyCollection<NamedOnnxValue> outputs,
+            RunOptions options)
+        {
+            if (inputNames.Count != inputValues.Count)
+            {
+                throw new ArgumentException($"Length of {nameof(inputNames)} ({inputNames.Count}) must match that of {nameof(inputValues)} ({inputValues.Count}).");
+            }
+
+            // prepare inputs
+            string[] inputNamesArray = inputNames as string[] ?? inputNames.ToArray();
+            IntPtr[] inputValuesArray = new IntPtr[inputNames.Count];
+            int inputIndex = 0;
+            foreach (var input in inputValues)
+            {
+                inputValuesArray[inputIndex] = input.Value;
+
+                inputIndex++;
+            }
+
+            // prepare outputs
+            var outputNamesArray = new string[outputs.Count];
+            var outputValuesArray = new IntPtr[outputs.Count];
+            var pinnedOutputBufferHandles = new System.Buffers.MemoryHandle[outputs.Count];
+
+            int outputIndex = 0;
+            foreach (var output in outputs)
+            {
+                outputNamesArray[outputIndex] = output.Name;
+
+                // create native OrtValue from the output if feasible, else throw notsupported exception for now
+                output.ToNativeOnnxValue(out outputValuesArray[outputIndex], out pinnedOutputBufferHandles[outputIndex]);
+
+                outputIndex++;
+            }
+
+            IntPtr status = NativeMethods.OrtRun(
+                                                _nativeHandle,
+                                                options.Handle,
+                                                inputNamesArray,
+                                                inputValuesArray,
+                                                (UIntPtr)inputNames.Count,
+                                                outputNamesArray,
+                                                (UIntPtr)outputs.Count,
+                                                outputValuesArray /* pointers to Pre-allocated OrtValue instances */
+                                                );
+
+            try
+            {
+                NativeApiStatus.VerifySuccess(status);
+            }
+            finally
+            {
+                // always unpin the output buffers, and delete the native Onnx value objects
+                for (int i = 0; i < outputs.Count; i++)
+                {
+                    NativeMethods.OrtReleaseValue(outputValuesArray[i]); // For elementary type Tensors, this should not release the buffer, but should delete the native tensor object.
+                                                                         // For string tensors, this releases the native memory allocated for the tensor, including the buffer
+                    pinnedOutputBufferHandles[i].Dispose();
```

Comment:
So the output buffer still contains the result of the (native) Run() but just gets unpinned (and the native OrtValue gets released along with it). Is this correct ?

Reply:
the output is a `NamedOnnxValue` object which is usually created from a `DenseTensor` with the pre-allocated buffer inside.